cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -337,10 +337,13 @@ struct SparkMurmurHash3_32 {
   template <typename TKey>
   result_type CUDA_DEVICE_CALLABLE compute(TKey const& key) const
   {
-    constexpr int len        = sizeof(TKey);
-    int8_t const* const data = reinterpret_cast<int8_t const*>(&key);
-    constexpr int nblocks    = len / 4;
+    return compute_bytes(reinterpret_cast<int8_t const*>(&key), sizeof(TKey));
+  }
 
+  result_type CUDA_DEVICE_CALLABLE compute_bytes(int8_t const* const data,
+                                                 cudf::size_type const len) const
+  {
+    int32_t const nblocks = len / 4;
     uint32_t h1           = m_seed;
     constexpr uint32_t c1 = 0xcc9e2d51;
     constexpr uint32_t c2 = 0x1b873593;
```

Comment:
It is unclear why `k1` is of type `uint32_t` in the first pass and `int32_t` in the second pass.

It appears that the remainder bytes are processed differently by Spark than the normal MurmurHash algorithm. This is clearer in the string implementation here: https://github.com/rapidsai/cudf/blob/0d5ec7f8902cbf3d5b73b037b9d0f5db66506d86/cpp/include/cudf/detail/utilities/hash_functions.cuh#L484
I updated the previous comment to clarify that the tail processing is Spark-specific.

This PR should update the string implementation to use the "compute bytes" function introduced in this PR since its behavior should be identical to a portion of the existing string implementation.

One possible question is whether we need to copy the `getblock32` function from the string implementation to `compute_bytes`. This would be helpful if that function generates aligned reads and the plain `reinterpret_cast` approach does not. This can be checked in the generated PTX/SASS. https://github.com/rapidsai/cudf/blob/0d5ec7f8902cbf3d5b73b037b9d0f5db66506d86/cpp/include/cudf/detail/utilities/hash_functions.cuh#L465-L469

Reply:
That reverse block processing is consistent with the non-spark version. Should that change too?

Reply:
I did some testing, there's rationale to the `RHS` of k1 being signed but not the `LHS`
```

    uint8_t u = 127; // 0x7f
    int8_t s = 127;
    cout << std::hex << (int32_t)u << endl;
    cout << std::hex << (uint32_t)u << endl;
    cout << std::hex << (int32_t)s << endl;
    cout << std::hex << (uint32_t)s << endl;

    uint8_t nu = -128; // 0x80
    int8_t ns = -128;
    cout << std::hex << (int32_t)nu << endl;
    cout << std::hex << (uint32_t)nu << endl;
    cout << std::hex << (int32_t)ns << endl;
    cout << std::hex << (uint32_t)ns << endl;
```

Results in:
```
7f
7f
7f
7f
80
80
ffffff80
ffffff80
```

Reply:
We will make a separate push for improving the implementations of the Murmur3 family of hash functions in a later PR.

Reply:
Started work on a followup PR for this and other changes targeted at 22.04 -- will link to this when I post but resolving this